src/utils/mtev_time.c

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright (c) 2016, Circonus, Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ *      notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ *      copyright notice, this list of conditions and the following
+ *      disclaimer in the documentation and/or other materials provided
+ *      with the distribution.
+ *    * Neither the name Circonus, Inc. nor the names of its contributors
+ *      may be used to endorse or promote products derived from this
+ *      software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <ck_pr.h>
+#include <math.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <time.h>
+#include <unistd.h>
+#include <mtev_time.h>
+#include <mtev_cpuid.h>
+#include <mtev_thread.h>
+#include <mtev_log.h>
+#include <mtev_defines.h>
+
+#ifdef __sun
+#include <sys/processor.h>
+#endif
+#if defined(linux) || defined(__linux) || defined(__linux__)
+#define _GNU_SOURCE
+#include <sched.h>
+
+#ifndef gettid
+static pid_t
+gettid(void)
+{
+  return syscall(__NR_gettid);
+}
+#endif /* gettid */
+#endif
+
+
+typedef uint64_t rdtsc_func(void);
+
+static __thread rdtsc_func *rdtsc_function;
+static __thread double ticks_per_nano;
+static __thread uint64_t last_ticks;
```

Comment:
I would create a
struct {
  rdtsc_func *rdtsc_function;
  double ticks_per_nano;
  uint64_t last_ticks;
  int32_t updating;
} coreclocks[NCPUS];

And then change static __thread stuff to just be 'int time_bound_cpu_id;'

then is last_ticks is too old, you try spinlock 'updating'... update and unlock.

This would mean that only one thread per core will have a hickup and if you care'd that much about hickups you could spin up dedicated bound maintenance threads they just force calibrate every 10s or so... then normal thread would always have "recently" calibrated clocks.